### ISSUE TITLE
Testnet Crashed `Q1.Z == alt_bn128_Fq2::one()'`

### ISSUE THREAD

**Issue by defuse** [zcash/zcash#437]
I logged in to the testnet host to update it, and I noticed that it crashed:

```
zcoin@zcointest:~$ cat testnet-data-dirs/0/stderr.log 
bitcoind: src/algebra/curves/alt_bn128/alt_bn128_pairing.cpp:347: libsnark::alt_bn128_ate_G2_precomp libsnark::alt_bn128_ate_precompute_G2(const libsnark::alt_bn128_G2&): Assertion `Q1.Z == alt_bn128_Fq2::one()' failed.
```

Looks like it happened in the verifier:

```
(leave) Call to zerocash_pour_ppzksnark_generator   [385.2429s x1.22]   (1447167782.5043s x0.00 from start)
(enter) Call to zerocash_pour_ppzksnark_verifier    [             ] (1447167783.9150s x0.00 from start)
  (enter) Call to zerocash_pour_input_map       [             ] (1447167783.9151s x0.00 from start)
  (leave) Call to zerocash_pour_input_map       [0.0002s x1.00] (1447167783.9153s x0.00 from start)
  (enter) Call to r1cs_ppzksnark_verifier_strong_IC [             ] (1447167783.9153s x0.00 from start)
    (enter) Call to r1cs_ppzksnark_verifier_process_vk  [             ] (1447167783.9153s x0.00 from start)
      (enter) Call to alt_bn128_ate_precompute_G2   [             ] (1447167783.9153s x0.00 from start)
      (leave) Call to alt_bn128_ate_precompute_G2   [0.0007s x1.00] (1447167783.9160s x0.00 from start)
      (enter) Call to alt_bn128_ate_precompute_G2   [             ] (1447167783.9160s x0.00 from start)
      (leave) Call to alt_bn128_ate_precompute_G2   [0.0007s x1.00] (1447167783.9167s x0.00 from start)
      (enter) Call to alt_bn128_ate_precompute_G1   [             ] (1447167783.9167s x0.00 from start)
      (leave) Call to alt_bn128_ate_precompute_G1   [0.0000s x1.04] (1447167783.9167s x0.00 from start)
      (enter) Call to alt_bn128_ate_precompute_G2   [             ] (1447167783.9167s x0.00 from start)
      (leave) Call to alt_bn128_ate_precompute_G2   [0.0007s x1.00] (1447167783.9174s x0.00 from start)
      (enter) Call to alt_bn128_ate_precompute_G2   [             ] (1447167783.9174s x0.00 from start)
      (leave) Call to alt_bn128_ate_precompute_G2   [0.0007s x1.00] (1447167783.9180s x0.00 from start)
      (enter) Call to alt_bn128_ate_precompute_G2   [             ] (1447167783.9180s x0.00 from start)
      (leave) Call to alt_bn128_ate_precompute_G2   [0.0007s x1.00] (1447167783.9187s x0.00 from start)
      (enter) Call to alt_bn128_ate_precompute_G1   [             ] (1447167783.9187s x0.00 from start)
      (leave) Call to alt_bn128_ate_precompute_G1   [0.0000s x1.02] (1447167783.9187s x0.00 from start)
      (enter) Call to alt_bn128_ate_precompute_G2   [             ] (1447167783.9187s x0.00 from start)
      (leave) Call to alt_bn128_ate_precompute_G2   [0.0007s x1.00] (1447167783.9194s x0.00 from start)
    (leave) Call to r1cs_ppzksnark_verifier_process_vk  [0.0041s x1.00] (1447167783.9194s x0.00 from start)
    (enter) Call to r1cs_ppzksnark_online_verifier_strong_IC    [             ] (1447167783.9194s x0.00 from start)
      (enter) Call to r1cs_ppzksnark_online_verifier_weak_IC    [             ] (1447167783.9194s x0.00 from start)
        (enter) Compute input-dependent part of A   [             ] (1447167783.9194s x0.00 from start)
        (leave) Compute input-dependent part of A   [0.0012s x1.00] (1447167783.9207s x0.00 from start)
        (enter) Check if the proof is well-formed   [             ] (1447167783.9207s x0.00 from start)
        (leave) Check if the proof is well-formed   [0.0000s x1.05] (1447167783.9207s x0.00 from start)
        (enter) Online pairing computations         [             ] (1447167783.9207s x0.00 from start)
          (enter) Check knowledge commitment for A is valid [             ] (1447167783.9207s x0.00 from start)
            (enter) Call to alt_bn128_ate_precompute_G1 [             ] (1447167783.9207s x0.00 from start)
            (leave) Call to alt_bn128_ate_precompute_G1 [0.0000s x1.06] (1447167783.9207s x0.00 from start)
            (enter) Call to alt_bn128_ate_precompute_G1 [             ] (1447167783.9207s x0.00 from start)
            (leave) Call to alt_bn128_ate_precompute_G1 [0.0000s x1.03] (1447167783.9207s x0.00 from start)
            (enter) Call to alt_bn128_ate_miller_loop   [             ] (1447167783.9207s x0.00 from start)
            (leave) Call to alt_bn128_ate_miller_loop   [0.0015s x1.00] (1447167783.9222s x0.00 from start)
            (enter) Call to alt_bn128_ate_miller_loop   [             ] (1447167783.9222s x0.00 from start)
            (leave) Call to alt_bn128_ate_miller_loop   [0.0015s x1.00] (1447167783.9238s x0.00 from start)
            (enter) Call to alt_bn128_final_exponentiation  [             ] (1447167783.9238s x0.00 from start)
              (enter) Call to alt_bn128_final_exponentiation_first_chunk    [             ] (1447167783.9238s x0.00 from start)
              (leave) Call to alt_bn128_final_exponentiation_first_chunk    [0.0001s x1.01] (1447167783.9238s x0.00 from start)
              (enter) Call to alt_bn128_final_exponentiation_last_chunk [             ] (1447167783.9239s x0.00 from start)
                (enter) Call to alt_bn128_exp_by_neg_z      [             ] (1447167783.9239s x0.00 from start)
                (leave) Call to alt_bn128_exp_by_neg_z      [0.0008s x1.00] (1447167783.9246s x0.00 from start)
                (enter) Call to alt_bn128_exp_by_neg_z      [             ] (1447167783.9247s x0.00 from start)
                (leave) Call to alt_bn128_exp_by_neg_z      [0.0008s x1.00] (1447167783.9254s x0.00 from start)
                (enter) Call to alt_bn128_exp_by_neg_z      [             ] (1447167783.9254s x0.00 from start)
                (leave) Call to alt_bn128_exp_by_neg_z      [0.0008s x1.00] (1447167783.9262s x0.00 from start)
              (leave) Call to alt_bn128_final_exponentiation_last_chunk [0.0025s x1.00] (1447167783.9263s x0.00 from start)
            (leave) Call to alt_bn128_final_exponentiation  [0.0026s x1.00] (1447167783.9263s x0.00 from start)
          (leave) Check knowledge commitment for A is valid [0.0057s x1.00] (1447167783.9263s x0.00 from start)
          (enter) Check knowledge commitment for B is valid [             ] (1447167783.9264s x0.00 from start)
            (enter) Call to alt_bn128_ate_precompute_G2 [             ] (1447167783.9264s x0.00 from start)
```

**Comment by daira** [comment]
I didn't see this bug before. It seems worrying since it could be a DoS vector. Do we still have that crashes/437 directory? Did we change that assertion to ``assert_except``?

**Comment by ebfull** [comment]
The assertion in this code basically is checking if the point is in the affine (z coordinate is 1 in the jacobian) during precomputation (basically, a part of the pairing that can be done with just the G2 element, because G2 arithmetic is so expensive). The point at infinity cannot be expressed in the affine. The likely cause was a proof or verification key that contained a spurious (or all) zeroes, which would have caused libsnark's deserialization to produce the point at infinity. That won't work in the pairing.

The assertion this ticket refers to was indeed changed to an exception. And I don't think you could trigger it, since you cannot deserialize the point at infinity in our implementation anymore.

**Comment by daira** [comment]
I'm satisfied that this is no longer exploitable.

**Comment by ebfull** [comment]
I think I found a tricky way you could trigger the bug, maybe.

You could coax vk<sub>x</sub> (the packed nullifiers and commitments and so on) into being the negative of a valid proof's π<sub>A</sub> during the QAP divisibility check. I don't think there's a way to coax it without beating ECDLP. The same thing appears in the same coefficient check. Can you take a look?

(By the way, the implementation in libsnark triggers the QAP divisibility check first, whereas in the paper I'm reading it's described as the final step, which again I don't think is a problem, but it does impact exploitability.)

**Comment by ebfull** [comment]
Again, in the worst case is should just throw an exception anyway, so I'm not suggesting this is actually "exploitable."

**Comment by daira** [comment]
Right, but let's fix this if we have time.

**Comment by daira** [comment]
Putting this in rc1 just to remind me to look at it, not because we need to fix it in rc1.

**Comment by ebfull** [comment]
@daira this got forgotten, can it be closed now?

**Comment by nathan-at-least** [comment]
Ping @daira to echo @ebfull's query: can this be safely closed now?

**Comment by ebfull** [comment]
I took some time today to analyze this so we could close the ticket.

**Summary**: There's no bug, but there are a few cool tests that could be added.

Upstream libsnark will assert when G2 points at infinity are provided to the pairing (precomputation). We changed this assertion to an exception, so it shouldn't be a problem. I further checked and I am confident there is no way to even trigger the exception because our zk-SNARK serialization code in Zcash rejects all points at infinity.

I reopened this ticket originally because I had come up with [a way](https://github.com/zcash/zcash/issues/437#issuecomment-245116065) to cause G1 points at infinity to be supplied to the pairing function using a trick. I am now sure the trick is possible, π<sub>A</sub>/π'<sub>A</sub> can be computed using the A query given a negation of the verifier's input. In order to actually do this, you need to access parts of the A query that libsnark's naive key generator will discard as a space optimization, but they can definitely be obtained from our MPC transcript. ([I wrote a very lousy test for this.](https://gist.github.com/ebfull/4cf9d6b9b0c088db490cea3c075f6f63))

This isn't a problem for G1 points though, libsnark handles this properly as this test (not yet in Zcash's test suite) shows:

```c++
TEST(proofs, g1_pairing_at_infinity)
{
    for (size_t i = 0; i < 100; i++) {
        auto r1 = curve_G1::random_element();
        auto r2 = curve_G2::random_element();
        ASSERT_TRUE(
            curve_pp::reduced_pairing(curve_G1::zero(), r2) ==
            curve_GT::one()
        );
        ASSERT_TRUE(
            curve_pp::final_exponentiation(
                curve_pp::double_miller_loop(
                    curve_pp::precompute_G1(curve_G1::zero()),
                    curve_pp::precompute_G2(r2),
                    curve_pp::precompute_G1(curve_G1::zero()),
                    curve_pp::precompute_G2(r2)
                )
            ) ==
            curve_GT::one()
        );
        ASSERT_TRUE(
            curve_pp::final_exponentiation(
                curve_pp::double_miller_loop(
                    curve_pp::precompute_G1(r1),
                    curve_pp::precompute_G2(r2),
                    curve_pp::precompute_G1(curve_G1::zero()),
                    curve_pp::precompute_G2(r2)
                )
            ) ==
            curve_pp::reduced_pairing(r1, r2)
        );
        ASSERT_TRUE(
            curve_pp::final_exponentiation(
                curve_pp::double_miller_loop(
                    curve_pp::precompute_G1(curve_G1::zero()),
                    curve_pp::precompute_G2(r2),
                    curve_pp::precompute_G1(r1),
                    curve_pp::precompute_G2(r2)
                )
            ) ==
            curve_pp::reduced_pairing(r1, r2)
        );
    }
}
```

We should also have a test for the G2 point at infinity precomputation exception. I also want to make a test that triggers a G1 point at infinity in the JoinSplit verifier for an (invalid) proof, but I'll have to grab parts of the MPC transcript to do that.

**Comment by nathan-at-least** [comment]
@ebfull IIUC, this sounds like three tests. Does it makes sense to have all in the Zcash test suite (prior to bringing libsnark in-tree, #2252) ? If so, does it seem possible to have a PR ready today for 1.0.9 or not?

**Comment by ebfull** [comment]
@nathan-at-least Only the test that I posted in the comment is likely to make it any time soon. It can be shoved into `proof_tests` in the gtest suite as-is. The others are too complicated and require either changes to libsnark or me digging through MPC code again.

**Comment by nathan-at-least** [comment]
Would you be available and interested to make a PR with the first test?

**Comment by ebfull** [comment]
@nathan-at-least https://github.com/zcash/zcash/pull/2399

**Comment by nathan-at-least** [comment]
We've merged #2399. It sounds like there are a few more tests we should add, and if those are added after #2252 it may be simpler.

**Comment by nathan-at-least** [comment]
I raised the priority of this issue, even if we're confident in the safety, adding tests to verify security is top priority.
